crates/shared/src/code_simulation.rs

```diff
@@ -178,6 +178,39 @@ impl TryFrom<StateOverride> for StateObject {
     }
 }
 
+/// A code simulator that uses Web3 in the general case, but will create and
+/// save failed simulations on Tenderly to facilitate debugging.
+pub struct Web3ThenTenderly {
```

Comment:
I'm of the opinion that there should be no `CodeSimulating` trait, there should only be a single code simulator type, and it should do "web3 then tenderly" by default. This would simplify the code. There's no reason to have three trait implementations just to facilitate a reasonable simulation behavior.

You probably already know this and are taking this route to stay consistent with the existing code style.